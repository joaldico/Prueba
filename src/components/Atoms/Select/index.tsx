import { MenuItem, TextField } from '@mui/material';

import { useStyles } from '../../../styles/global/inputStyles';

type ChangeEventHandler = React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;

interface SelectProps {
  id?: string;
  label?: string;
  value?: string;
  onChange?: ChangeEventHandler;
  data: any;
  fullWidth?: boolean;
  placeholder?: string;
}

function Select(props: SelectProps): JSX.Element {
  const classes = useStyles();

  return (
    <TextField
      id={props.id}
      select
      label={props.label}
      variant="outlined"
      value={props.value}
      onChange={props.onChange}
      fullWidth={props.fullWidth}
      InputLabelProps={{
        shrink: true,
        classes: {
          asterisk: classes.asteriskColor,
          root: classes.labelRoot,
          focused: classes.cssFocused,
        },
      }}
      SelectProps={{
        displayEmpty: true,
      }}
      InputProps={{
        className: classes.input,
      }}
      size="small"
    >
      {props.placeholder && (
        <MenuItem
          value={''}
          disabled
          classes={{
            selected: classes.selected,
            root: classes.rootMenu,
          }}
        >
          {props.placeholder}
        </MenuItem>
      )}

      {props.data.map((option: any) => {
        return (
          <MenuItem
            key={option.value}
            value={option.value}
            classes={{
              selected: classes.selected,
              root: classes.root,
            }}
          >
            {option.label}
          </MenuItem>
        );
      })}
    </TextField>
  );
}

export default Select;
