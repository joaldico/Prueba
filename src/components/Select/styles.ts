import { Theme } from '@mui/material/styles';
import { makeStyles } from '@mui/styles';

export const useStyles = makeStyles((theme: Theme) => ({
  rootPlaceholder: {
    fontStyle: 'italic',
    '&:hover': {
      backgroundColor: `${theme.palette.primary.light} !important`,
    },
  },
  root: {
    '&:hover': {
      backgroundColor: `${theme.palette.primary.light} !important`,
    },
  },
}));
