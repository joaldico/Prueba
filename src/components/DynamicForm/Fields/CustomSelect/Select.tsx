import { IndeterminateCheckBox } from '@mui/icons-material';
import CheckBoxIcon from '@mui/icons-material/CheckBox';
import CheckBoxOutlineBlankIcon from '@mui/icons-material/CheckBoxOutlineBlank';
import { TextField } from '@mui/material';
import Autocomplete from '@mui/material/Autocomplete';
import Checkbox from '@mui/material/Checkbox';
import ListItemText from '@mui/material/ListItemText';
import React, { FC } from 'react';
import { v4 as uuidv4 } from 'uuid';

import { inputPropsDefault, useStylesInputLabel } from '../styles';

const icon = <CheckBoxOutlineBlankIcon fontSize="small" />;
const checkedIcon = <CheckBoxIcon fontSize="small" />;
const IndeterminateIcon = <IndeterminateCheckBox fontSize="small" color="primary" />;

type SelectComponentProps = {
  props?: any;
  id?: any;
  value?: any;
  disableCloseOnSelect?: any;
  noWrap?: any;
  disableClearable?: any;
  limitTags?: any;
  selectAll?: any;
  placeholder?: any;
  type?: any;
  trigger?: any;
  error?: any;
  onChange?: any;
  label?: any;
  data?: any;
  required?: any;
  fullWidth?: any;
  disabled?: any;
  multiple?: any;
};

const SelectComponent: FC<SelectComponentProps> = ({
  id,
  label,
  value,
  onChange,
  error,
  data,
  required,
  fullWidth,
  disabled,
  multiple,
  disableCloseOnSelect,
  disableClearable,
  limitTags,
  selectAll,
  placeholder,
  type,
  trigger,
}) => {
  const getSortedData = () => {
    return data.sort((x: any, y: any) =>
      x.label.localeCompare(y.label, 'es', { sensitivity: 'base', numeric: true })
    );
  };

  //Opcion de seleccionar todos
  const allOption = {
    label: 'Seleccionar todos',
    value: 'select-all',
  };

  const classesInput = useStylesInputLabel();
  return (
    <Autocomplete
      id={id}
      multiple={multiple}
      limitTags={limitTags}
      options={data?.length > 0 ? getSortedData() : []}
      disableCloseOnSelect={disableCloseOnSelect}
      disableClearable={disableClearable}
      disabled={disabled}
      onChange={(event, newValue) => {
        //Si la opcion que seleccionan es "seleccionar todos"
        if (selectAll) {
          if (newValue.find((option: any) => option.value === allOption.value)) {
            //Verificamos si estÃ¡n todas las opciones seleccionadas
            if (value.length != data.length) {
              onChange(data);
            } else {
              //Si estan todas seleccionadas limpiamos los seleccionados
              onChange([]);
            }
          } else {
            onChange(newValue);
          }
        } else {
          onChange(newValue);
        }
        trigger(id);
        return newValue;
      }}
      value={value}
      getOptionLabel={(option) => option.label}
      renderOption={(option: any, { selected }) => {
        const selectedOptionCheckbox = option.value == allOption.value;
        return (
          <React.Fragment>
            {multiple &&
              (selectedOptionCheckbox ? (
                <Checkbox
                  checked={value.length === data.length}
                  style={{ marginRight: 8 }}
                  indeterminate={value.length > 0 && value.length < data.length}
                  // onChange={handleSelectAll}
                  indeterminateIcon={IndeterminateIcon}
                  checkedIcon={checkedIcon}
                  icon={icon}
                />
              ) : (
                <Checkbox
                  icon={icon}
                  checkedIcon={checkedIcon}
                  style={{ marginRight: 8 }}
                  checked={selected}
                />
              ))}
            <ListItemText primary={option.label} />
          </React.Fragment>
        );
      }}
      style={{ width: '100%' }}
      renderInput={(params) => (
        <TextField
          {...params}
          label={label}
          variant="outlined"
          disabled={disabled}
          error={!!error}
          helperText={error ? error.message : null}
          required={required}
          fullWidth={fullWidth}
          value={value}
          type={type}
          InputLabelProps={{
            ...inputPropsDefault(classesInput),
          }}
          placeholder={placeholder}
        />
      )}
    />
  );
};

SelectComponent.defaultProps = {
  id: uuidv4(),
  value: [],
  disableCloseOnSelect: true,
  noWrap: false,
  disableClearable: false,
  limitTags: -1,
  selectAll: false,
  placeholder: 'Seleccione',
  type: 'text',
};

export default SelectComponent;
