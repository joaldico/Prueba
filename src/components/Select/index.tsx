import { MenuItem, TextField } from '@mui/material';
import classNames from 'classnames';

import { useStyles as inputStyles } from '../../styles/global/inputStyles';
import { useStyles } from './styles';

type ChangeEventHandler = React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;

interface SelectProps {
  id?: string;
  label?: string;
  value?: string;
  error?: any;
  onChange?: ChangeEventHandler;
  data: any;
  fullWidth?: boolean;
  placeholder?: string;
  className?: any;
  MenuClass?: any;
  required?: boolean;
  disabled?: boolean;
  disabledPlaceholder?: boolean;
  textCleanValue: string;
  variant?: any;
}

function Select(props: SelectProps): JSX.Element {
  const classes = useStyles();
  const classesInputGlobalStyles = inputStyles();
  return (
    <TextField
      id={props.id}
      select
      defaultValue=""
      label={props.label}
      variant={props.variant ? props.variant : 'outlined'}
      value={props.value || ''}
      error={props.error ? props.error : false}
      onChange={props.onChange}
      fullWidth={props.fullWidth}
      InputLabelProps={{
        shrink: true,
        classes: {
          asterisk: classesInputGlobalStyles.asteriskColor,
          root: classesInputGlobalStyles.labelRoot,
          focused: classesInputGlobalStyles.cssFocused,
        },
      }}
      SelectProps={{
        displayEmpty: true,
      }}
      InputProps={{
        className: classNames(props.className, classesInputGlobalStyles.input),
      }}
      required={props.required || false}
      disabled={props.disabled}
      helperText={
        <>
          <span className={classesInputGlobalStyles.p}>
            {props.error ? props.error.message : null}
          </span>
        </>
      }
    >
      {props.placeholder && (
        <MenuItem
          value={''}
          disabled={props.value == '' ? props.disabledPlaceholder : false}
          classes={{
            selected: classNames(classesInputGlobalStyles.selected, props.MenuClass),
            root: classNames(classes.rootPlaceholder, props.MenuClass),
          }}
        >
          {props.value == '' ? props.placeholder : props.textCleanValue}
        </MenuItem>
      )}

      {props.data.map((option: any) => {
        return (
          <MenuItem
            key={option.value}
            value={option.value}
            classes={{
              selected: classNames(classesInputGlobalStyles.selected, props.MenuClass),
              root: classNames(classes.root, props.MenuClass),
            }}
          >
            {option.label}
          </MenuItem>
        );
      })}
    </TextField>
  );
}

Select.defaultProps = {
  disabledPlaceholder: true,
  textCleanValue: 'Limpiar busqueda',
  variant: 'outlined',
};

export default Select;
