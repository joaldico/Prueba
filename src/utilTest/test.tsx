import { createTheme, PaletteMode, PaletteOptions } from '@mui/material';
import { ThemeProvider as MuiThemeProvider } from '@mui/styles';
import React from 'react';

import { COLOR_DEFAULT } from '../constants/constants';

export const ThemeProviderTest: React.FC = ({ children }) => {
  const mainPalette: PaletteOptions = {
    mode: 'light' as PaletteMode,
    primary: {
      main: COLOR_DEFAULT,
      light: COLOR_DEFAULT,
    },
  };

  const customMaterialTheme = createTheme({
    palette: mainPalette,
    breakpoints: {
      values: {
        xs: 0,
        sm: 768,
        md: 1024,
        lg: 1440,
        xl: 1920,
      },
    },
  });

  return <MuiThemeProvider theme={customMaterialTheme}>{children}</MuiThemeProvider>;
};
