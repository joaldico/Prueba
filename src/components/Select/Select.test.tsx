import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import Select from './index';
import { ThemeProviderTest } from '../../utilTest/test';

describe('Select Snapshot ', () => {
  it('Snapshot Select default', () => {
    const props = {
      id: 'select_id',
      label: 'label',
      value: '1',
      onChange: () => {},
      data: [{ label: 'data 1', value: '1' }],
      fullWidth: true,
      placeholder: 'placeholder',
    };

    const { asFragment } = render(
      <ThemeProviderTest>
        <Select {...props} />
      </ThemeProviderTest>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
