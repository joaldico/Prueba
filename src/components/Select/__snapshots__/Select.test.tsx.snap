// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Snapshot  Snapshot Select default 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="select_id"
      id="select_id-label"
    >
      label
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl makeStyles-input-3 css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        aria-describedby="select_id-helper-text"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="select_id-label select_id"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="select_id"
        role="button"
        tabindex="0"
      >
        data 1
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value="1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1z7n62"
        >
          <span>
            label
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
      id="select_id-helper-text"
    >
      <span
        class="makeStyles-p-5"
      />
    </p>
  </div>
</DocumentFragment>
`;
